### PR TITLE
[PFD] Improve Vfe and Vfen

### PR DESCRIPTION
**Pull request by MisterChocker**
## Summary of Changes

- Fixed wrong VFE and VFENEXT
- Added table to speed util

List of correct speeds:
| Config | knots IAE |
| ------ | ----- |
| 1 | 230 |
| 1 + F | 215 |
| 2 | 200 |
| 3 | 185 |
| 4 | 177 |

## References

image

## Additional context

Before, VFE and VFE next for FLAPS 1 and FLAPS 1 + F were not correct and displayed wrong.
Not VFE and VFE next for FLAPS 1 and FLAPS 1 + F are displayed with the correct logic.

Discord username (if different from GitHub): DerL30N#3751

## Testing instructions

Check Vfe / Vmo and Vfen for every flaps config.
Also test for auto retract flaps from config 1 + F to 1.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js [hunk 1]
```diff
@@ -36,21 +41,24 @@ class A32NX_Speeds {
...
 
             SimVar.SetSimVarValue("L:A32NX_SPEEDS_VS", "number", speeds.vs);
             SimVar.SetSimVarValue("L:A32NX_SPEEDS_VLS", "number", speeds.vls);
             SimVar.SetSimVarValue("L:A32NX_SPEEDS_F", "number", speeds.f);
             SimVar.SetSimVarValue("L:A32NX_SPEEDS_S", "number", speeds.s);
             SimVar.SetSimVarValue("L:A32NX_SPEEDS_GD", "number", speeds.gd);
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFE", "number", speeds.vfe);
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFEN", "number", speeds.vfeN);
+
+            console.log("VFE: " + speeds.vfe + ", VFEN: " + speeds.vfeN);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js:61]: Remove `console.log` calls

**Review by Benjozork — APPROVED** [review]
LGTM

**Comment by Lazarus320** [comment]
Category: QA Tester
Name: Lazarus#4286
Date of testing: 26/12/2020
Version of the sim: 1.12.13
PR Tested: #2722 
Object of test: Improve vfe and vfe plus flap auto retraction. 
Testing technique: Tested the new speed and auto flap retraction doing multiple takeoffs/approaches. 
Steps you've made to test the Pull Request:
1. The auto retraction to flaps 1 from 1 + F happens when the aircraft is very heavy where the minimum speed to go from configuration 1+F to zero can reach or exceed 215kts (around 213 in this test at almost max tow) and since there is no way to go from Config 1+F to Config 1 by action on the flap lever  the flaps will automatically retract at 210kts, putting the aircraft into configuration 1 you can notice below the flaps automatically retract but the slats stay in position. 

https://user-images.githubusercontent.com/74972083/103147565-ccfe9b00-4791-11eb-8804-3de60afaa810.mp4

1. Flaps 1 vfe checked. 
image


2. Flaps 1 + F vfe checked. 
image

3. Flaps 2 vfe checked. 
image

4. Flaps 3 vfe checked. 
image

5. Vmo in clean config checked. 
image

6. Vfen for flaps 1 checked. 
image
 
7. Vfen for flaps 2 checked. 
image
 
8. Vfen for flaps 3 checked. 
image

Flight notes (If required): N/A

Issues (If Any): N/A
Overall Rating: LGTM
Conclusions of the testing: Everything seems good well done.

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js (hunk 1 above)
Comment (MisterChocker) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js:61]:
```suggestion
```

**Comment by donstim** [comment]
This overrides the VFE in the flight model cfg file, right?  Because I fixed the wrong VFE for 1+F in the flight model cfg file in #2698. Also, #2698 removes the auto-retraction of 1+F to 1 to fix the uncontrollable stall and dive problem. I hope this PR does not reenable it!

**Comment by MisterChocker** [comment]
@donstim No, this pr adds a table to our speed calculations, just like the config file to ensure the correct VFE and VFEN is displayed and used for master caution (overspeed). I've chosen this approach to ensure the use of correct speeds at all time.
Btw. before VFE and VFEN for flaps 1 or 1 + F were not displayed correctly.

**Comment by sidnov** [comment]
Category: QA Tester
Name: sidnov#8337
Date of testing: 27/12/2020
Version of the sim: 1.12.13.0
PR Tested: #2722
Object of test: Vfe and Vfen speeds, flap auto-retraction check

Testing technique: Straight-out departure from RWY

Steps you've made to test the Pull Request:
1. Spawned at VABB RWY27
2. Set FCU to manual mode with initial speed 200kts, runway heading 270, altitude 4000ft
3. After takeoff, engaged autopilot, checked speed and set altitude to V/S mode to climb at 1400ft/min
4. Increased speed to 215kts, observed auto-retraction of flaps around 210kts and slats retained their extended position
5. Retracted flaps completely, continued climb to 4000ft. After reaching 4000ft, reduced speed to 200kts.
6. Extended Flaps to config 1, checked and verified Vfe=230kts Vfen=200kts. Reduced speed to 180kts
7. Extended Flaps to config 2, checked and verified Vfe=200kts Vfen=185kts. Reduced speed to 160kts
8. Extended Flaps to config 3, checked and verified Vfe=185kts, then to config 4 Vfe=177kts.

F1+F
F1 (land)
F2 (land)
F3 (land)
FFull (land)



Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: Flaps retract to 0 automatically after crossing 210kts IAE, while slats remain in position. The max speeds (Vfe) for each flap config was verified. The predictive next config (Vfen) speeds were checked too. Screenshots attached.

**Comment by iamdefective1** [comment]
Category: QA Tester
Name: IAMDEFECTIVE1#9293
Date of testing: 27/12/2020 
Version of the sim: 1.12.13.0
PR Tested: #2722
Object of test: Vmax, Vfe and Vfen speeds on PFD. 1+f auto retraction.

Testing technique: EGCC 23r Circuit, noted V speeds while on approach
Steps you've made to test the Pull Request:
1. Depart 23r, and make a right hand departure out.
2. Approach into airport, take note of Vfen, Vfe and Vmax speed.

Flight notes:
Flaps Clean VMax - 350KT
Flaps 1 VFen - 230KT
Flaps 1+F Vfe - 210KT
Flaps 1 Vfe - 230KT
Flaps 2 Vfen - 200KT
Flaps 2 Vfe - 205KT
Flaps 3 Vfen - 185KT
Flaps 3 Vfe - 185KT
Flaps Full Vfen - 177KT
Flaps Full Vfe - 177KT

Screenshots:
Flaps Clean Vmax
image

Flaps 1 Vfen
image

Flaps 1+F Vfe
image

Flaps 1 Vfe
image

Flaps 2 Vfen
image

Flaps 2 Vfe
image

Flaps 3 Vfen
image

Flaps 3 Vfe
image

Flaps Full Vfen
image

Flaps Full Vfe
image



Issues (If Any): (None) **
Overall Rating: Good
Conclusions of the testing: Vfe and Vfen speeds show up on the PFD, attatched SS's

**Flaps 1+F auto-retraction had failed when I tried, but others had reported it okay.

----------------------
This is my first QA test so please leave me some feedback (dms is okay)!
Thanks

**Pull request merged**